### PR TITLE
feat!: use `eslint-webpack-plugin`

### LINKED ISSUE #38: [v3] Use `eslint-webpack-plugin`
Use `eslint-webpack-plugin` instead of `eslint-loader`

### PR DESCRIPTION
**Pull request by ricardogobbosouza**
Resolve #38 

BREAKING CHANGE:
- Enabled `cache` by default
- Enabled `lintDirtyModulesOnly` by default
- Use `eslint-webpack-plugin` instead of `eslint-loader` deprecated.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -57,89 +57,113 @@ export default {
...
+Specify directories, files, or globs. Must be relative to `options.context`.
+Directories are traveresed recursively looking for files matching `options.extensions`.
+File and glob patterns ignore `options.extensions`.
 
 ### `extensions`
 
-- Type: `Array[String]`
-- Default: `['ts', 'js', 'vue']`
+- Type: `String|Array[String]`
+- Default: `'js'`
```
Comment (rchl) [inline README.md:101]: Isn't the default `['ts', 'js', 'vue']`?

File: README.md [hunk 2]
```diff
@@ -57,89 +57,113 @@ export default {
...
 
-This option will enable caching of the linting results into a file. This is particularly useful in reducing linting time when doing a full build.
+**Note**: The cache is enabled by default to decrease execution time.
 
-This can either be a `boolean` value or the cache directory path(ex: `'./.eslint-loader-cache'`).
+### `context`
 
-If `cache: true` is used, the cache is written to the `./node_modules/.cache/eslint-loader` directory. This is the recommended usage.
+- Type: `String`
+- Default: `compiler.context`
```
Comment (rchl) [inline README.md:78]: I'm not sure what `compiler.context` means. Should it something like "Nuxt srcDir" instead?

File: README.md [hunk 3]
```diff
@@ -57,89 +57,113 @@ export default {
 
 ## Options
 
-You can pass [eslint options](http://eslint.org/docs/developer-guide/nodejs-api#cliengine).
+You can pass [eslint options](https://eslint.org/docs/developer-guide/nodejs-api#%E2%97%86-new-eslint-options).
```
Comment (rchl) [inline README.md:60]: The correct link seems to be: https://eslint.org/docs/developer-guide/nodejs-api#-new-eslintoptions

File: README.md [hunk 4]
```diff
@@ -57,89 +57,113 @@ export default {
...
 
-**Note**: That the config option you provide will be passed to the `CLIEngine`. This is a different set of options than what you'd specify in `package.json` or `.eslintrc`. See the [eslint docs](http://eslint.org/docs/developer-guide/nodejs-api#cliengine) for more detail.
+Note that the config option you provide will be passed to the `ESLint` class.
+This is a different set of options than what you'd specify in `package.json` or `.eslintrc`.
+See the [eslint docs](https://eslint.org/docs/developer-guide/nodejs-api#%E2%97%86-new-eslint-options) for more details.
```
Comment (rchl) [inline README.md:64]: Fix link.

File: README.md [hunk 5]
```diff
@@ -57,89 +57,113 @@ export default {
...
+
+A string indicating the root of your files.
 
 ### `eslintPath`
 
 - Type: `String`
 - Default: `eslint`
 
-Path to `eslint` instance that will be used for linting. If the `eslintPath` is a folder like a official eslint, or specify a `formatter` option. Now you dont have to install `eslint`.
+Path to `eslint` instance that will be used for linting. If the `eslintPath` is a folder like a official eslint, or specify a `formatter` option. now you dont have to install `eslint`.
```
Comment (rchl) [inline README.md:87]: This paragraph is a bit broken grammatically. Can you rephrase it?

File: README.md [hunk 6]
```diff
@@ -57,89 +57,113 @@ export default {
...
+
+### `files`
+
+- Type: `String|Array[String]`
+- Default: `'.'`
+
+Specify directories, files, or globs. Must be relative to `options.context`.
+Directories are traveresed recursively looking for files matching `options.extensions`.
```
Comment (rchl) [inline README.md:95]: "traversed"

File: README.md [hunk 7]
```diff
@@ -57,89 +57,113 @@ export default {
...
-Extensions that will be used by the loader.
+Specify extensions that should be checked.
 
 ### `fix`
 
 - Type: `Boolean`
 - Default: `false`
 
-This option will enable [ESLint autofix feature](http://eslint.org/docs/user-guide/command-line-interface#fix).
+Will enable [ESLint autofix feature](https://eslint.org/docs/developer-guide/nodejs-api#%E2%97%86-eslint-outputfixes-results).
```
Comment (rchl) [inline README.md:110]: Correct link seems to be: https://eslint.org/docs/developer-guide/nodejs-api#cliengineoutputfixes

File: README.md [hunk 8]
```diff
@@ -57,89 +57,113 @@ export default {
...
 - Type: `String|Function`
-- Default: `stylish`
+- Default: `'stylish'`
+
+Accepts a function that will have one argument: an array of eslint messages (object). The function must return the output as a string. You can use official [eslint formatters](https://eslint.org/docs/user-guide/formatters/).
+
+### `lintDirtyModulesOnly`
+
+- Type: `Boolean`
+- Default: `false`
```
Comment (pi0) [inline README.md:124]: What if we enable this by default to speedup initial build and avoid unnecessary warnings that are not related to one's changes?

File: README.md (hunk 1 above)
Comment (ricardogobbosouza) [reply README.md]: Fixed. Thanks for review

File: README.md (hunk 2 above)
Comment (ricardogobbosouza) [reply README.md]: Fixed. Thanks for review

File: README.md (hunk 3 above)
Comment (ricardogobbosouza) [reply README.md]: Fixed. Thanks for review

File: README.md (hunk 4 above)
Comment (ricardogobbosouza) [reply README.md]: Fixed. Thanks for review

File: README.md (hunk 5 above)
Comment (ricardogobbosouza) [reply README.md]: Fixed. Thanks for review

File: README.md (hunk 6 above)
Comment (ricardogobbosouza) [reply README.md]: Fixed. Thanks for review

File: README.md (hunk 7 above)
Comment (ricardogobbosouza) [reply README.md]: Fixed. Thanks for review

File: README.md (hunk 8 above)
Comment (ricardogobbosouza) [reply README.md]: Sure. Enabled 😄

**Review by pi0 — APPROVED** [review]

**Comment by zoxon** [comment]
Do you test it before merging? Looks like module completely broken

**Comment by rlam3** [comment]
> Do you test it before merging? Looks like module completely broken

mines breaking too... idk where the upgrade guide is for this module... have you found a resolution or a work around?

**Comment by toddheslin** [comment]
Looks to be a problem with the cache. If pass:
```js
eslint: {
    cache: false,
  },
```
In my `nuxt.config.js` then it won't generate the `.eslintcache` file which seems to be causing the trouble.

I like the idea of the cache but something has gone wrong through the upgrade.

**Comment by rlam3** [comment]
@zoxon @toddheslin I did a quick doc for this upgrade. I believe upgrading from v2 to v3 was causing the problem. 
IDK if you are facing the same problem I was. But I got it to work after configuring the nuxt.config.js with eslint-webpack-plugin. It certainly was lacking in documentation.

https://aznric3boi91.medium.com/update-your-nuxtjs-eslint-module-from-v2-to-v3-20b5f0bcdb68

**Comment by toddheslin** [comment]
Awesome @rlam3 thanks for the article. Just a note for your 'after': you already destructured `isDev` and `isClient` so it should be (without `ctx`):
```js
export default {
	// Run ESLint on save
  build: {
    extend(config, { isDev, isClient }) {
      if (isDev && isClient) {
        const options = {
          exclude: ['node_modules']
        }
        const EslintPlugin = require('eslint-webpack-plugin')
        config.plugins.push(new EslintPlugin(options))
    }
  }
}
```
